elftools/common/construct_utils.py

```diff
@@ -47,49 +47,43 @@ def _build(self, obj, stream, context):
     def _sizeof(self, context):
         raise SizeofError("can't calculate size")
 
+class ULEB128(Construct):
+    """A construct based parser for ULEB128 encoding.
 
-def _LEB128_reader():
-    """ Read LEB128 variable-length data from the stream. The data is terminated
-        by a byte with 0 in its highest bit.
+       Incompatible with Python 2 - assumes that the return of read()
+       is an indexed collection of numbers.
     """
-    return RepeatUntil(
-                lambda obj, ctx: ord(obj) < 0x80,
-                Field(None, 1))
-
-
-class _ULEB128Adapter(Adapter):
-    """ An adapter for ULEB128, given a sequence of bytes in a sub-construct.
-    """
-    def _decode(self, obj, context):
+    def _parse(self, stream, context):
         value = 0
-        for b in reversed(obj):
-            value = (value << 7) + (ord(b) & 0x7F)
-        return value
-
-
-class _SLEB128Adapter(Adapter):
-    """ An adapter for SLEB128, given a sequence of bytes in a sub-construct.
+        shift = 0
+        while True:
+            data = stream.read(1)
+            if len(data) != 1:
+                raise FieldError("expected 1, found 0")
```

Comment:
Here and below: maybe something about premature end of stream instead? Like "stream ended in the middle of a LEB128 encoded value"?

Reply:
Good point. I was going for a drop-in replacement for the existing logic, including the exception wording. Will fix.